### PR TITLE
Refactoring shortest path/direct shortest path queries

### PR DESCRIPTION
**Pull request by TheMarex**
This pull request addresses https://github.com/Project-OSRM/osrm-backend/issues/1546. In the process I had to structurally change `shortest_path` and `direct_shortest_path` and used it as an opportunity for a refactor.

/cc @danpat for code review

[Review on Reviewable](https://reviewable.io/reviews/project-osrm/osrm-backend/1795)

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[plugins/viaroute.hpp, line 123 \[r1\]](https://reviewable.io:443/reviews/project-osrm/osrm-backend/1795#-K4YBE-K73_a9ERUJrLK:-K4YBE-K73_a9ERUJrLL:-804150890) ([raw file](https://github.com/project-osrm/osrm-backend/blob/e4d60d70d20f46d207d82499d6653e353e892482/plugins/viaroute.hpp#L123)):*
No need for constructing an expensive stream, use

    const auto thisIsAStdString = "hello world" + std::to_string(integral);

I think the whole file has this problem.

---

*[routing_algorithms/routing_base.hpp, line 194 \[r1\]](https://reviewable.io:443/reviews/project-osrm/osrm-backend/1795#-K4YCCEaceB29BNnixg-:-K4YCCEaceB29BNnixg0:1250862421) ([raw file](https://github.com/project-osrm/osrm-backend/blob/e4d60d70d20f46d207d82499d6653e353e892482/routing_algorithms/routing_base.hpp#L194)):*
If this really is a model of the `RandomIterator` concept, there is no need for `std::prev` or similar, you can just use `--` and `++`.

Maybe `static_assert` on the iterator tag via iterator traits.

---

*[routing_algorithms/shortest_path.hpp, line 129 \[r1\]](https://reviewable.io:443/reviews/project-osrm/osrm-backend/1795#-K4YDwxX4qUBPbeUeZL8:-K4YDwxX4qUBPbeUeZL9:1660723191) ([raw file](https://github.com/project-osrm/osrm-backend/blob/e4d60d70d20f46d207d82499d6653e353e892482/routing_algorithms/shortest_path.hpp#L129)):*
what
why
what is this monster

can you handle 14 parameters, let alone the`2^4=16` possible states from the 4 bools alone --- because I can't

---

*[routing_algorithms/shortest_path.hpp, line 231 \[r1\]](https://reviewable.io:443/reviews/project-osrm/osrm-backend/1795#-K4YEWLYKZTFOflAk05E:-K4YEWLYKZTFOflAk05F:2101406251) ([raw file](https://github.com/project-osrm/osrm-backend/blob/e4d60d70d20f46d207d82499d6653e353e892482/routing_algorithms/shortest_path.hpp#L231)):*
see above

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/project-osrm/osrm-backend/1795)*

**Comment by TheMarex** [comment]
*[routing_algorithms/routing_base.hpp, line 194 \[r1\]](https://reviewable.io:443/reviews/project-osrm/osrm-backend/1795#-K4YCCEaceB29BNnixg-:-K4YGObJ2Vr8c45SHsQh:-850313398) ([raw file](https://github.com/project-osrm/osrm-backend/blob/e4d60d70d20f46d207d82499d6653e353e892482/routing_algorithms/routing_base.hpp#L194)):*
Doesn't that map to the same thing anyway?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/project-osrm/osrm-backend/1795)*

File: routing_algorithms/direct_shortest_path.hpp [hunk 1]
```diff
@@ -64,143 +65,71 @@ class DirectShortestPathRouting final
...
-                super::RoutingStep(forward_core_heap, reverse_core_heap, &middle, &distance,
-                                   min_edge_offset, true, false);
-            }
-            if (!reverse_core_heap.Empty())
-            {
-                super::RoutingStep(reverse_core_heap, forward_core_heap, &middle, &distance,
-                                   min_edge_offset, false, false);
-            }
+            super::Search(forward_heap, reverse_heap, distance, packed_leg);
+            std::copy(packed_leg.begin(), packed_leg.end(), std::ostream_iterator<NodeID>(std::cout, ", "));
```
Comment (TheMarex) [inline routing_algorithms/direct_shortest_path.hpp]: Foo forgot this debug output.

File: routing_algorithms/shortest_path.hpp [hunk 2]
```diff
@@ -51,294 +53,482 @@ class ShortestPathRouting final
...
+                const int total_distance_to_forward,
+                const int total_distance_to_reverse,
+                int &new_total_distance_to_forward,
+                int &new_total_distance_to_reverse,
+                std::vector<NodeID> &leg_packed_path_forward,
+                std::vector<NodeID> &leg_packed_path_reverse) const
+    {
+        // silence unused warning
+        (void) search_from_forward_node;
+        (void) search_from_reverse_node;
```
Comment (danpat) [inline routing_algorithms/shortest_path.hpp]: Might be neater to do this in the signature instead:

    const bool /* search_from_forward_node */,
    const bool /* search_from_reverse_node */,

Why are these unused?

Comment (TheMarex) [inline routing_algorithms/shortest_path.hpp]: They are used, see the assertions below.

Comment (danpat) [inline routing_algorithms/shortest_path.hpp]: They're only used during debug builds, but they're polluting the function signature under normal use.  I think the assertion might belong at the call site for `SearchLoop`, there only seems to be one.

**Comment by TheMarex** [comment]
Merged after fixing the issues.

**Pull request merged**